flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -0,0 +1,268 @@
+			.entrySet()
+			.stream()
+			.flatMap(entry -> allocateLogicalSlotsFromSharedSlot(sharedSlotProfileRetriever, entry.getKey(), entry.getValue()))
+			.collect(Collectors.toMap(SlotExecutionVertexAssignment::getExecutionVertexId, a -> a));
+
+		return executionVertexIds.stream().map(assignments::get).collect(Collectors.toList());
+	}
+
+	@Override
+	public void cancel(ExecutionVertexID executionVertexId) {
+		ExecutionSlotSharingGroup executionSlotSharingGroup =
+			slotSharingStrategy.getExecutionSlotSharingGroup(executionVertexId);
+		Preconditions.checkNotNull(
+			executionSlotSharingGroup,
+			"There is no ExecutionSlotSharingGroup for ExecutionVertexID " + executionVertexId);
+		SharedSlot slot = sharedSlots.get(executionSlotSharingGroup);
+		if (slot != null) {
+			slot.cancelLogicalSlotRequest(executionVertexId);
+		} else {
+			LOG.debug("There is no slot for ExecutionSlotSharingGroup of ExecutionVertexID {}", executionVertexId);
+		}
+	}
+
+	private Stream<? extends SlotExecutionVertexAssignment> allocateLogicalSlotsFromSharedSlot(
+			SharedSlotProfileRetriever sharedSlotProfileRetriever,
+			ExecutionSlotSharingGroup executionSlotSharingGroup,
+			Collection<ExecutionVertexID> executions) {
+		SharedSlot sharedSlot = getOrAllocateSharedSlot(executionSlotSharingGroup, sharedSlotProfileRetriever);
+		return executions.stream().map(sharedSlot::createAssignment);
+	}
+
+	private SharedSlot getOrAllocateSharedSlot(
+			ExecutionSlotSharingGroup executionSlotSharingGroup,
+			SharedSlotProfileRetriever sharedSlotProfileRetriever) {
+		return sharedSlots
+			.computeIfAbsent(executionSlotSharingGroup, group -> {
+				SlotRequestId physicalSlotRequestId = new SlotRequestId();
+				CompletableFuture<PhysicalSlot> physicalSlotFuture = sharedSlotProfileRetriever
+					.getSlotProfileFuture(group)
+					.thenCompose(slotProfile -> slotProvider.allocatePhysicalSlot(
+						new PhysicalSlotRequest(physicalSlotRequestId, slotProfile, slotWillBeOccupiedIndefinitely)))
+					.thenApply(PhysicalSlotRequest.Result::getPhysicalSlot);
+				return new SharedSlot(physicalSlotRequestId, group, physicalSlotFuture);
+			});
+	}
+
+	private class SharedSlot implements SlotOwner, PhysicalSlot.Payload {
+		private final SlotRequestId physicalSlotRequestId;
+
+		private final ExecutionSlotSharingGroup executionSlotSharingGroup;
+
+		private final CompletableFuture<PhysicalSlot> slotContextFuture;
+
+		private final DualKeyLinkedMap<ExecutionVertexID, SlotRequestId, SlotExecutionVertexAssignment> requestedLogicalSlots;
+
+		private SharedSlot(
+				SlotRequestId physicalSlotRequestId,
+				ExecutionSlotSharingGroup executionSlotSharingGroup,
+				CompletableFuture<PhysicalSlot> slotContextFuture) {
+			this.physicalSlotRequestId = physicalSlotRequestId;
+			this.executionSlotSharingGroup = executionSlotSharingGroup;
+			this.slotContextFuture = slotContextFuture.thenApply(physicalSlot -> {
```

Comment:
I think we need to cancel all the pending logical slot futures and remove this shared slot if the allocation fails.

Reply:
Sorry I made a mistake here. Failed physical slot allocation will result in failed logical slot future, and finally removes the shared slot itself.